drv/lpc55-sprot-server/src/handler.rs

```diff
@@ -50,213 +54,125 @@ impl Handler {
-                        tx_buf.serialize(MsgType::UpdPrepImageUpdateRsp, rsp)
-                    }
-                    Err(e) => Err((tx_buf, e.into())),
-                }
-            }
-            MsgType::UpdWriteOneBlockReq => {
-                match hubpack::deserialize::<u32>(rx_payload) {
-                    Ok((block_num, block)) => {
-                        let rsp: UpdateRspHeader = self
-                            .update
-                            .write_one_block(block_num as usize, block)
-                            .map(|_| None)
-                            .map_err(|e| e.into());
-
-                        tx_buf.serialize(MsgType::UpdWriteOneBlockRsp, rsp)
-                    }
-                    Err(e) => Err((tx_buf, e.into())),
-                }
-            }
-
-            MsgType::UpdAbortUpdateReq => {
-                let rsp: UpdateRspHeader = self
-                    .update
-                    .abort_update()
-                    .map(|_| None)
-                    .map_err(|e| e.into());
-                tx_buf.serialize(MsgType::UpdAbortUpdateRsp, rsp)
-            }
-            MsgType::UpdFinishImageUpdateReq => {
-                let rsp: UpdateRspHeader = self
-                    .update
-                    .finish_image_update()
-                    .map(|_| None)
-                    .map_err(|e| e.into());
-                tx_buf.serialize(MsgType::UpdFinishImageUpdateRsp, rsp)
-            }
-            MsgType::SinkReq => {
-                // The first two bytes of a SinkReq payload are the U16
-                // mod 2^16 sequence number.
-                if rx_payload.len() >= 2 {
-                    let tx_payload = tx_buf.payload_mut();
-                    tx_payload[..2].copy_from_slice(&rx_payload[..2]);
-                    tx_buf.from_existing(MsgType::SinkRsp, 2)
-                } else {
-                    Ok(tx_buf.no_payload(MsgType::SinkRsp))
-                }
-            }
-            MsgType::DumpReq => {
-                let addr =
-                    u32::from_le_bytes(rx_payload[0..4].try_into().unwrap());
+            ReqBody::IoStats => Ok(RspBody::IoStats(stats.clone())),
+            ReqBody::Sprockets(req) => {
+                // TODO: Don't unwrap!
+                Ok(RspBody::Sprockets(
+                    self.sprocket.handle_deserialized(req).unwrap_lite(),
+                ))
+            }
+            ReqBody::Dump { addr } => {
                 ringbuf_entry!(Trace::Dump(addr));
-
                 let dumper = Dumper::from(DUMPER.get_task_id());
-
-                let rval: u32 = if let Err(e) = dumper.dump(addr) {
-                    e.into()
-                } else {
-                    0
-                };
-
-                let tx_payload = tx_buf.payload_mut();
-                tx_payload[0..4].copy_from_slice(&rval.to_le_bytes());
-                tx_buf.from_existing(MsgType::DumpRsp, 4)
+                match dumper.dump(addr) {
```

Comment:
FWIW, this could also be written as

```rust
dumper.dump(addr).map_err(SprotError::Dump)?;
Ok(RspBody::Ok)
```

I like the concise version better, but use whichever you prefer.

Reply:
I always forget I can map a variant as a function. Thanks!

Reply:
Fixed.